plugins/catalog-react/src/components/EntityKindPicker/EntityKindPicker.tsx

```diff
@@ -109,18 +52,26 @@ function useEntityKindFilter(opts: { initialFilter: string }): {
     }
   }, [filters.kind]);
 
-  const { availableKinds, loading, error } = useAvailableKinds();
-
   useEffect(() => {
     updateFilters({
       kind: selectedKind ? new EntityKindFilter(selectedKind) : undefined,
     });
   }, [selectedKind, updateFilters]);
 
+  // Set selected kinds on query parameter updates; this happens at initial page load and from
+  // external updates to the page location.
```

Comment:
You probably do not want to do this. There are situations where it's very easy to get into infinite loops where the component sets the filter state, which updates the internal filters and mirrors it into the query, which the component then picks up as slightly different (maybe some casing difference or whatnot), and then goes back to update the filter state again ... etc.

Query state should most likely ONLY be taken into account in the initial setup on first render, e.g. what the header does https://github.com/backstage/backstage/blob/2fb5e6e3ba64fdc5c3f23f34a5afaef76ed89b27/plugins/catalog/src/components/CatalogKindHeader/CatalogKindHeader.tsx#L82

Reply:
Yeah! It works great without monitoring the query.

Reply:
Mm. Hmmm. Now that I think about it some more, I think there was something about users clicking in-page links that led to the same page but with different query parameters - and then, the filters didn't change to reflect that. So the code probably was there to handle that. But the thing is - it'd have to lead to a rerender for that hook to run anyway, right? So that makes me wonder if the only way that we could possibly deal with that correctly, is to add an actual event handler either on the location or on react-router. But alright, I digress. If this seemingly works, then it works, my musings be damned. So let's leave the check that you had in place.

Reply:
> So let's leave the check that you had in place.

Just so I don't misunderstand your comment. You want me to bring back this code I removed:
```
  // Set selected kinds on query parameter updates; this happens at initial page load and from
  // external updates to the page location.
  useEffect(() => {
    if (flattenedQueryKind) {
      setSelectedKind(flattenedQueryKind);
    }
  }, [flattenedQueryKind]);
```

If I'm not mistaken, it will most likely triggers extra renders in some cases. But perhaps that's the price we have to pay?

Reply:
Yeah. This code has never probably never worked as intended. The tests I removed issues a rerender, hence why the useEffect() actually register the changed query parameter. That is not really a reflection of the real world, is it?

If we just update the location through `window.history.pushState()`the useEffect won't work. Though I'm really a Backend developer, so I'm not sure if I'm missing anything. I'm keeping the useEffect() on queryKind removed. It should not really work today, so if there is a need, it should probably be implemented with a location listener or something.

Reply:
Alright. Yeah to be clear, I'm in the "do whatever works" camp here since this isn't "my" code and I'm slightly concerned as well that maybe it wasn't doing the right thing in the first place. I'm just happy you're looking at it and making things better. :)

Reply:
My mistake. After some extensive testing, the `SidebarSubmenuItems`:
```
            <SidebarSubmenuItem
              title="Domains"
              to="catalog?filters[kind]=domain"
              icon={DomainIcon}
            />
```
actually needs this code to work:
```
  // Set selected kinds on query parameter updates; this happens at initial page load and from
  // external updates to the page location.
  useEffect(() => {
    if (flattenedQueryKind) {
      setSelectedKind(flattenedQueryKind);
    }
  }, [flattenedQueryKind]);
```

I brought back the testcases and state updates for the queryParam.